File: NiceReads/Backend/Server/tos-server/src/application/models/SkillLevel.ts
import { Document } from './Document';
import { Uuid } from '../../common/uuid';
import { IEqualible } from '../../common/IEqualible';
import { IHashCodable } from '../../common/IHashCodable';
import { StringUtils } from '../../common/StringUtils';

export class SkillLevel extends Document implements IEqualible, IHashCodable {
    constructor(
        public name: string = '',
        public description: string ='',
        public order: number = -1,
        id: Uuid = Uuid.EMPTY
    ) {
        super(id);
    }

    equals(obj: any): boolean {
        const skillLevel: SkillLevel = obj as SkillLevel;
        return !!skillLevel && skillLevel.name === this.name;
    }

    getHashCode(): number {
        if (StringUtils.isEmptyOrWhiteSpace(this.name)) {
            return 0;
        }
        return StringUtils.hashCode(this.name);
    }
}

File: EngagementHub/TOS.EngagementHub.Models/SkillLevel.cs
namespace TOS.EngagementHub.Models
{
    public class SkillLevel : EngagementModel
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public int Order { get; set; }

        public override bool Equals(object obj)
        {
            SkillLevel other = obj as SkillLevel;
            return other != null && other.Name != null && other.Name.Equals(this.Name);
        }

        public override int GetHashCode()
        {
            return (Name ?? string.Empty).GetHashCode();
        }

        public static bool operator ==(SkillLevel first, SkillLevel second)
        {
            return first != null && first.Equals(second);
        }

        public static bool operator !=(SkillLevel first, SkillLevel second)
        {
            return !(first.Order == second.Order);
        }

        public static bool operator >(SkillLevel first, SkillLevel second)
        {
            return first.Order > second.Order;
        }

        public static bool operator <(SkillLevel first, SkillLevel second)
        {
            return first.Order < second.Order;
        }

        public static bool operator >=(SkillLevel first, SkillLevel second)
        {
            return first == second || first.Order > second.Order;
        }

        public static bool operator <=(SkillLevel first, SkillLevel second)
        {
            return first == second || first.Order < second.Order;
        }
    }
}
